bin/faucet/src/server/pow.rs

```diff
@@ -100,6 +201,10 @@ pub(crate) fn check_pow_solution(seed: &str, solution: u64) -> Result<(), Invali
         return Err(InvalidRequest::InvalidPoW);
     }
 
+    // If we get here, the solution is valid
+    // Remove the challenge to prevent reuse
+    challenge_state.remove_challenge(seed);
```

Comment:
This confuses me as I had assumed this struct was holding an index of created (and/or completed) challenges. So I don't follow why it would be removed upon success?

Reply:
I added this removal for the challenge to be invalid from this point, without needed to keep any other field or state. Since it would no longer be in the challenge cache, any new request using the challenge would be rejected.

Reply:
Ah OK this is a different model to what I had assumed. I was thinking of keeping a map of used challenges to prevent replay.

What is stopping someone from replaying a challenge within its time range?

Reply:
As in, is it possible to have multiple concurrent runs of `pub fn check_pow_solution(` for the same challenge to succeed?

Reply:
you're right, there is a small gap between checking the challenge validity and removing it from the cache that allows for challenges to be be duplicated. I'm moving the lock.